cluster/src/main/java/org/apache/iotdb/cluster/metadata/CMManager.java

```diff
@@ -1298,6 +1299,49 @@ protected MeasurementMNode getMeasurementMNode(MNode deviceMNode, String measure
     return super.showTimeseries(plan, context);
   }
 
+  public Set<PartialPath> getLocalDevices(ShowDevicesPlan plan) throws MetadataException {
+    return super.getDevices(plan);
+  }
+
+  @Override
+  public Set<PartialPath> getDevices(ShowDevicesPlan plan) throws MetadataException {
+    ConcurrentSkipListSet<PartialPath> resultSet = new ConcurrentSkipListSet<>();
+    ExecutorService pool = new ThreadPoolExecutor(THREAD_POOL_SIZE, THREAD_POOL_SIZE, 0,
+        TimeUnit.SECONDS, new LinkedBlockingDeque<>());
+    List<PartitionGroup> globalGroups = metaGroupMember.getPartitionTable().getGlobalGroups();
+
+    int limit = plan.getLimit() == 0 ? Integer.MAX_VALUE : plan.getLimit();
+    int offset = plan.getOffset();
+    // do not use limit and offset in sub-queries unless offset is 0, otherwise the results are
+    // not combinable
```

Comment:
Have we check the limit and offset is reasonable , for example both is greater than 0?

Reply:
It should be checked in plan generation, instead of execution.